File: C_Sharp/Framework/TransactionScript.cs
namespace Framework
{
    public interface ITransactionScriptExecution
    {
        void RunScript();
    }

    public interface ITransactionScript : ITransactionScriptExecution
    {
        IUnitOfWork CreateUnitOfWork();
        void PreExecuteBody();
        
        void PostExecuteBody();
    }

    public interface ITransactionScript<TInput, TOutput> : ITransactionScript
    {
        TInput Input { get; set; }
        TOutput Output { get; set; }
        TOutput ExecutionBody();
    }

    public abstract class TransactionScript<TInput, TOutput> : ITransactionScript<TInput, TOutput>
    {
        public TInput Input { get; set; }
        public TOutput Output { get; set; }

        protected IRepositoryRegistry RepositoryRegistry { get; private set; }
        protected IMapperRegistry MapperRegistry { get; private set; }

        protected TransactionScript(IRepositoryRegistry repositoryRegistry, IMapperRegistry mapperRegistry)
        {
            RepositoryRegistry = repositoryRegistry;
            MapperRegistry = mapperRegistry;
        }

        public virtual void PreExecuteBody()
        {
        }

        public abstract TOutput ExecutionBody();

        public virtual void PostExecuteBody()
        {
        }

        public void RunScript()
        {
            PreExecuteBody();

            Output = ExecutionBody();

            PostExecuteBody();
        }

        public IUnitOfWork CreateUnitOfWork()
        {
            return new UnitOfWork();
        }
    }
}

File: Android/framework/src/main/java/com/TransactionScript.java
package com;

import com.Interfaces.ITransactionScriptConcrete;
import com.Interfaces.IUnitOfWork;
import com.Interfaces.Registries.IMapperRegistry;
import com.Interfaces.Registries.IRepositoryRegistry;

/**
 * Created by aiko on 4/1/17.
 */

public abstract class TransactionScript<TInput, TOutput> implements ITransactionScriptConcrete<TInput, TOutput> {
        TInput _input;

        public TInput GetInput()
        {
        return _input;
        }

        public void SetInput(TInput input)
        {
            _input = input;
        }

    TOutput _output;

    public TOutput GetOutput()
    {
     return _output;
    }

    public void SetOutput(TOutput output)
    {
        _output = output;
    }

    IRepositoryRegistry _repositoryRegistry;
    IMapperRegistry _mapperRegistry;

    public IRepositoryRegistry GetRepositoryRegistry(){return _repositoryRegistry;}
    public IMapperRegistry GetMapperRegistry(){return _mapperRegistry;}

    protected TransactionScript(IRepositoryRegistry repositoryRegistry, IMapperRegistry mapperRegistry)
    {
        _repositoryRegistry = repositoryRegistry;
        _mapperRegistry = mapperRegistry;
    }

    public void PreExecuteBody(){}

    public abstract TOutput ExecutionBody();

    public void PostExecuteBody(){}

    public void RunScript()
    {
        PreExecuteBody();

        _output = ExecutionBody();

        PostExecuteBody();
    }

    public IUnitOfWork CreateUnitOfWork()
    {
        return null;
    }
}
